File: LeetCode/Java/src/solutions/p0208_ImplementTrie/P0208.java
package solutions.p0208_ImplementTrie;

public class P0208 {
    /*
     * Solution v1: Brute Force
     *
     * Runtime: 38 ms @ (beats) 63.80%
     * Memory Usage: 50.9 MB @ (beats) 85.74%
     */
    class Trie {

        Object[] root;

        public Trie() {
            this.root = new Object[27];
        }

        public void insert(String word) {
            Object[] node = this.root;

            for (int i = 0; i < word.length(); i++) {
                int j = word.charAt(i) - 'a';
                if (node[j] == null) {
                    node[j] = new Object[27];
                }

                node = (Object[]) node[j];

            }

            node[26] = Boolean.TRUE;

        }

        public boolean search(String word) {
            Object[] node = this.root;

            for (int i = 0; i < word.length(); i++) {
                int j = word.charAt(i) - 'a';
                if (node[j] == null){
                    return false;
                }
                node = (Object[]) node[j];
            }

            return node[26] == Boolean.TRUE;
        }

        public boolean startsWith(String prefix) {
            Object[] node = this.root;

            for (int i = 0; i < prefix.length(); i++) {
                int j = prefix.charAt(i) - 'a';

                if (node[j] == null){
                    return false;
                }
                node = (Object[]) node[j];
            }

            return true;
        }
    }
}

File: LeetCode/Python/solutions/p0208.py
#
# Solution v1: Brute Force
#
# Runtime: 180 ms @ (beats) 57.25%
# Memory Usage: 31.5 MB @ (beats) 74.7%
#
# class TrieNode:
#     def __init__(self):
#         self.children = {}
#         self.is_end = False
#
#
# class Trie:
#
#     def __init__(self):
#         self.root = TrieNode()
#
#     def insert(self, word: str) -> None:
#         node = self.root
#         for c in word:
#             if c not in node.children:
#                 node.children[c] = TrieNode()
#             node = node.children[c]
#
#         node.is_end = True
#
#     def search(self, word: str) -> bool:
#         node = self.root
#         i = 0
#         while i < len(word):
#             if word[i] in node.children.keys():
#                 node = node.children[word[i]]
#                 i += 1
#             else:
#                 return False
#
#         return node.is_end
#
#     def startsWith(self, prefix: str) -> bool:
#         node = self.root
#         i = 0
#         while i < len(prefix):
#             if prefix[i] in node.children.keys():
#                 node = node.children[prefix[i]]
#                 i += 1
#             else:
#                 return False
#
#         return True

#
# Solution v1.1: Brute Force
# Simplify the object structure
#
# Runtime: 127 ms @ (beats) 93.71%
# Memory Usage: 27.5 MB @ (beats) 92.85%
#
class Trie:

    def __init__(self):
        self.root = {}

    def insert(self, word: str) -> None:
        node = self.root
        for c in word:
            if c not in node:
                node[c] = {}
            node = node[c]

        node["*"] = ""

    def search(self, word: str) -> bool:
        node = self.root
        i = 0
        while i < len(word):
            c = word[i]
            if c in node:
                node = node[c]
                i += 1
            else:
                return False

        return "*" in node

    def startsWith(self, prefix: str) -> bool:
        node = self.root
        for p in prefix:
            if p in node:
                node = node[p]
            else:
                return False

        return True


if __name__ == "__main__":
    trie = Trie()
    print(f"insert apple: {trie.insert('apple')}")  # None
    print(f"search apple: {trie.search('apple')}")  # True
    print(f"search app: {trie.search('app')}")  # False
    print(f"startsWith app: {trie.startsWith('app')}")  # Ture
    print(f"insert app: {trie.insert('app')}")  # None
    print(f"search app: {trie.search('app')}")  # True
